sensor/common/context/context_provider.go

```diff
@@ -0,0 +1,94 @@
+package contextprovider
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/pkg/centralsensor"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/sensor/common"
+	"github.com/stackrox/rox/sensor/common/message"
+)
+
+// ContextProvider Sensor component that provides a common context.
+// This context will be cancelled and reset if Sensor enters Offline mode.
+type ContextProvider interface {
+	common.SensorComponent
+	GetContext() context.Context
+}
+
+var _ ContextProvider = (*contextProviderImpl)(nil)
+
+// NewContextProvider initializes a new ContextProvider
+func NewContextProvider() ContextProvider {
+	ret := &contextProviderImpl{
+		firstCentralConnection: &atomic.Bool{},
+	}
+	// We initialize here and not in Start to avoid races with other sensor components
+	// that might call GetContext on their respective Start functions.
+	ret.init(func() (context.Context, func()) {
+		return context.WithCancel(context.Background())
+	})
+	return ret
+}
+
+type contextProviderImpl struct {
+	mu              sync.RWMutex
+	cancelContextFn func()
```

Comment:
Could we maybe use the original type here?

```suggestion
	cancelContextFn context.CancelFunc
```